c10/core/Device.cpp

```diff
@@ -73,12 +73,17 @@ Device::Device(const std::string& device_string) : Device(Type::CPU) {
     type_ = parse_type(s);
 
     std::string device_index = device_string.substr(index + 1);
+    std::size_t pos = 0;
```

Comment:
@SsnL would it be more elegant to do this with std::regex? https://github.com/pytorch/pytorch/blob/5ec87a3c1a6acfb0eb4704873649ffcd7e65c221/c10/core/Device.cpp#L41-L63. It looks like the minimum version of gcc we support is 5 so this should work now.

Reply:
To make regex work, we would have to account for future/other device types (ROCM?, etc.). I wasn't sure how to do that properly... So I kept `stoi`.

Reply:
You could do something with regex groups, like the following so that we don't hardcode cpu & cuda:
```
#include <iostream>
#include <regex>

int main()
{
    static const std::regex with_idx_regex("([a-zA-Z_]+):([0-9]+)");
    static const std::regex without_idx_regex("([a-zA-Z_]+)");

    std::smatch match;
    const std::string device_string = "cuda:1";
    const bool ok = std::regex_match(device_string, match, with_idx_regex);
    if (ok) {
        std::cout << "device: " << match[1].str() << std::endl;
        std::cout << "idx: " << match[2].str() << std::endl;
    }   
}
```
The regex-based approach will probably have different error messages from the current parsing based approach, but it's a lot nicer than having to maintain our own parser for this

Reply:
using regex now!

Reply:
friendly ping :) @zou3519

Reply:
Thanks! I'll take a look by eod tomorrow